File: numTUVrxn.f90
PROGRAM numTUVrxn


!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~!
! Variable declarations:                                               !
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~!

  IMPLICIT NONE

  INTEGER        :: i,ierr,nrxn
  CHARACTER(80)  :: line,ifile
  CHARACTER(1)   :: reset

!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~!

! Retrieve file name from first programme argument,
! otherwise ask for it:
  CALL getarg(1,ifile)
  IF (ifile == ' ') THEN
    WRITE(*,"('Name of input file: ',A)",advance='no')
    READ(*,*) ifile
  ENDIF
  ifile = TRIM(ADJUSTL(ifile))

! retrieve choice of reaction switch from 2nd programme argument
  CALL getarg(2,reset)

! Open input file and temporary output file:
  OPEN(11,file=ifile)
  OPEN(12,FILE='ofile.txt')

!––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––!

! Read away header and spectral weighting functions:
  head: DO WHILE (line /= '===== Available photolysis reactions:')
    READ(11,'(A)') line
    WRITE(12,'(A)') trim(line)
  ENDDO head

!––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––!

! Initialise count of output (true) reactions:
  i = 0
  nrxn = 0

! Loop over photolysis reactions:
  phot: DO
! Increase counter for total number of reactions
    i = i + 1
! Read in photolysis reactions line by line
    READ(11,'(A)') line

! Exit condition of loop when reaching line starting with '===':
    IF(line(:3) == '===') THEN
! Write final line to temporary output file:
      WRITE(12,'(A)') trim(line)
! Reset rxn counter to total number of reactions, if all reactions are
! forced to true:
      IF(reset=='T') THEN
        nrxn = i-1
! Reset rxn counter to 0, if all reactions are forced to false:
       ELSEIF(reset=='F') THEN
        nrxn = 0
      ENDIF
!     Exit loop:
      EXIT
    ENDIF

! Overwrite reaction switched, if forced to true or false:
    IF(reset=="T" .or. reset=="t") THEN
      WRITE(line(1:1),'(A1)') "T"
     ELSEIF(reset=="F" .or. reset=="f") THEN
      WRITE(line(1:1),'(A1)') "F"
     ELSE
! Otherwise count number of true reactions:
      IF(line(:1)=='T') nrxn = nrxn + 1
    ENDIF
! Write each reaction line with new reaction numbers to temporary output
    WRITE(line(2:4),'(I3)') i
    WRITE(12,'(A)') trim(line)
  ENDDO phot

!––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––!

! Adjust parameter nmj

! Re-read temporary output file from the beginning
! and print output to another temporary output file with adjusted parameter nmj

! Rewind ofile.txt to top
  REWIND(12)
! Open 2nd temporary output file
  OPEN(13,FILE='ofile.dat')
! Loop over first output file
  ll: DO
!   Read again line by line
    READ(12,'(A)',IOSTAT=ierr) line
! Exit on end of file
    IF(ierr<0) THEN
      EXIT
     ELSEIF(ierr>0) THEN
      STOP "Read error in temporary output file 'ofile.txt'!"
    ENDIF
! Override number of output (true) reactions
    IF(INDEX(line,'nmj')>0) WRITE(line(61:66),'(I6)') nrxn
! Write (adjusted) lines to 2nd temporary file
    WRITE(13,'(A)') trim(line)
  ENDDO ll

!––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––!

! Close all files
  CLOSE(11)
  CLOSE(12,STATUS='DELETE')
  CLOSE(13)

! Use Unix to move temporary output over original TUV input file:
  CALL SYSTEM('mv ofile.dat '//trim(ifile))

END PROGRAM numTUVrxn

File: numTUVrxn.jl
#!/usr/local/bin/julia


"""
# Module numTUVrxn

Correct reaction numbers after alterations of the TUV input file.
Preserve flags for output or set/unset all.
"""
module numTUVrxn

# Define location of external self-made modules
# (Add or modify to include your own directories)
# Local Mac:
if isdir("/Applications/bin/data/jl.mod") &&
  all(LOAD_PATH.!="/Applications/bin/data/jl.mod")
  push!(LOAD_PATH,"/Applications/bin/data/jl.mod")
end
# earth0:
if isdir(joinpath(homedir(),"Util/auxdata/jl.mod")) &&
  all(LOAD_PATH.!=joinpath(homedir(),"Util/auxdata/jl.mod"))
  push!(LOAD_PATH,joinpath(homedir(),"Util/auxdata/jl.mod"))
end
import fhandle: test_file, rdfil

# Append ARGS by empty strings
# to avoid error messages in case of missing input
for i = 1:2-length(ARGS)  push!(ARGS," ")  end


################################################################################
### Functions:                                                                 #
################################################################################


"""
    get_rxn(lines)

From lines in TUV input file, derive top section (top), flags for reactions (flags)
and reaction strings (rxns) from reaction section, and bottom line (bottom)
and return separately.
"""
function get_rxn(lines)

  # Initialise
  rfl = false # flag for reaction section
  flags = String[]
  rxns  = String[]
  top   = String[]
  bottom = String

  # Loop over lines
  for i in eachindex(lines)
    # On keyword for reactions section, save last line to top and set flag for rxn section
    if lines[i] == "===== Available photolysis reactions:"
      push!(top,lines[i]); rfl = true
    # Save last line of file to bottom
    elseif rfl==true && lines[i][1:3]=="==="  bottom = lines[i]
    # In reaction section, only save flags for reactions and reaction strings
    elseif rfl==true  push!(flags,lines[i][1:1]); push!(rxns,lines[i][6:end])
    # Save header and spectral weighting functions to top
    else   push!(top,lines[i])
    end
  end

  # return saved sections separately
  return top, bottom, flags, rxns
end #function find_rxn


"""
    renumber(fl,flags,rxns,top,bottom)

Overwrite reaction numbers in TUV input file and optionally flags for reactions
and adjust parameter `nmj` for number of flagged reactions.

If fl is set to T/t or F/f in the second script argument all reaction flags are
set to T or F, respectively. The reaction section is derived from the reaction flags
flags (or the overwritten flags), new consecutive reaction numbers and the reaction
labels (rxns). All sections top, the new reaction section and bottom are concatenated
and returned as array (lines).
"""
function renumber(fl,flags,rxns,top,bottom)

  # Set all flags to T, if fl = T,t,True,... and set number of flags to # of rxns
  if lowercase(fl[1])=='t'  flags .= "T"; trxn = length(flags)
  # Set all flags to F, if fl = F,f,False,... and set number of flags to 0
  elseif lowercase(fl[1])=='f'  flags .= "F"; trxn = 0
  # fl not specified, leave flags as they are
  elseif fl==" "  trxn = count(f->f=="T",flags)
  # if fl is file path, determine flags from md file
  else
    flags .= "F"
    open(fl,"r") do f
      for line in eachline(f)
        if length(matchall(r"J\(",line))≥2
          (nr,crxn) = strip.(split(line," | ")[3:4]); nr = parse(Int64,nr)
          if rxns[nr] ≠ crxn
            println("\033[95mWarning reaction labels in md file not identical to TUV input file for reaction $nr.\033[0m")
          end
          flags[nr] = "T"
        end
      end
    end
    trxn = count(f->f=="T",flags)
  end

  # Rewrite reaction section with consecutive reaction numbers starting at 1
  rlines = String[]
  for i = eachindex(flags)
    push!(rlines,@sprintf("%s%3d %s", flags[i],i,rxns[i]))
  end

  # Search for parameter nmj and replace with new number of true flags
  idx = 0
  # Loop over top section
  for i = eachindex(top)
    # Find nmj
    try idx = search(top[i],"nmj = ")[end] end
    if idx > 0
      srch = search(top[i],r"[0-9]+",idx)
      repl = string(trxn)
      # Adjust length of new nmj to length of old nmj, if shorter
      try  repl = " "^(length(srch)-length(repl))*repl  end
      # replace
      top[i] = top[i][1:srch[end]-length(repl)]*repl*top[i][srch[end]+1:end]
      break
    end
  end

  # return array with complete file content of concatenated sections
  lines = vcat(top,rlines,bottom)

  return lines
end #function renumber


"""
    wrtfile(file,lines)

Overwrite TUV input file (file) with adjusted lines (lines) bearing new reaction
numbers and optionally reaction flags.
"""
function wrtfile(file,lines)

  # remove old TUV file and write new content to file with same file name
  rm(file); open(file,"w") do f
    [println(f, lines[i]) for i = eachindex(lines)]
  end
end #function wrtfile


################################################################################
### End of Functions; Main script:                                             #
################################################################################

# Get file name from script argument or user input and read content (lines)
ifile = ARGS[1]
if splitdir(ifile)[1] == "" && basename(pwd()) == "numTUVrxn"  ifile = joinpath("..",ifile)  end
ifile = test_file(ifile)
lines = rdfil(ifile)
# Split content in sections and get flags
# and reaction strings from reaction section
top, bottom, flags, rxns = get_rxn(lines)
# Overwrite reaction numbers and optionally flags
# and return complete and revised file content
lines = renumber(ARGS[2],flags,rxns,top,bottom)
# Overwrite old TUV input file
wrtfile(ifile,lines)

end #module numTUVrxn
